apps/sms/views/shared/js/activity_handler.js

```diff
@@ -21,22 +23,27 @@ const ActivityDataType = {
 };
 
 var ActivityHandler = {
-  // Will hold current activity object
-  _activity: null,
+  _activityClient: null,
 
   init: function() {
     if (!window.navigator.mozSetMessageHandler) {
       return;
     }
 
-    // A mapping of MozActivity names to their associated event handler
-    window.navigator.mozSetMessageHandler('activity',
-      this._onActivity.bind(this, {
-        'new': this._onNewActivity,
-        'share': this._onShareActivity
-      })
+    this._activityClient = new ActivityClient();
+
+    this._activityClient.on(
+      'new-activity-request', this._onNewActivity.bind(this)
+    );
+
+    this._activityClient.on(
+      'share-activity-request', this._onShareActivity.bind(this)
     );
 
+    // Currently both client and service for activity resides in the same
+    // context, so we instantiate both.
+    (new ActivityService());
```

Comment:
where will the service live later on ? in a worker ? in the main view ?

I'm afraid this won't work like this... the "threads" library is supposed to spawn the service itself, either in a worker, in a shared worker or in window (which is a hidden iframe in the main view). But for the activity, we actually want the service to live in the main view itself so that we can handle the system message... at least until the various bugs around system messages are fixed.

So either we can augment the "threads" lib to handle this case, or we'll need some sort of adhoc mechanism (using broadcast channel ?).